### PR TITLE
Resolve frontend conflicts

### PR DESCRIPTION
**Pull request by ccasin**
A few notes:
- Most conflicts are due to the "move modes to the parsetree" PR being merged.  I took the current flambda-backend code for these, generally.
- There are some cases where code was duplicated by the merge (so you'll see some cases where it will look like this PR just deletes some random code or just deletes a conflict without taking either side).
- something weird happened with `printer/printast_with_mappings.ml` - it doesn't seem to have been merged (possibly only things under `ocaml/` were merged, and this is just the one that fails due to parsetree changes?).  I just manually diffed it with the tip of flambda-backend and integrated the needed changes.
- The addition of `ocaml/parsing/parser_types.ml` required changes in the dynlink dune file - I'm not sure why these weren't needed when it was the lrgrep PR was merged, but the dynlink dune file in this branch is pretty different from the one in flambda-backend.

Request review from @ncik-roberts

### CODE REVIEW

File: ocaml/parsing/pprintast.ml [hunk 1]
```diff
@@ -992,47 +911,6 @@ and sugar_expr ctxt f e =
...
-    match params with
-    | [] -> ()
-    | _ :: _ -> pp f "%a@;" (list (function_param ctxt) ~sep:"@ ") params
-  in
-  pp f "%t%a%s@;%a"
-    pp_params
-    (option (function_constraint ctxt) ~first:"@;") constraint_
-    delimiter
-    (function_body (under_functionrhs ctxt)) body
-
```
Comment (ncik-roberts) [inline ocaml/parsing/pprintast.ml]: I suggest keeping this hunk and deleting the `flambda-backend/main` one (that should be close to identical). Reasoning: I moved this hunk here in the merge to minimize diff with upstream.

File: ocaml/parsing/ast_invariants.ml [hunk 2]
```diff
@@ -39,18 +39,9 @@ let module_type_substitution_missing_rhs loc =
...
-||||||| a198127529
-let no_val_params loc = err loc "Functions must have a value parameter."
-
-let non_jane_syntax_function loc =
-  err loc "Functions must be constructed using Jane Street syntax."
-=======
   err loc "Functions must have a value parameter."
 let redundant_nested_constraints loc =
   err loc "Nested pattern constraints must all specify a type"
->>>>>>> flambda-backend/main
```
Comment (ncik-roberts) [inline ocaml/parsing/ast_invariants.ml]: I suggest keeping the `err loc "Function without any value parameters"` portion from HEAD to minimize diff with upstream.

File: ocaml/parsing/parser.mly [hunk 3]
```diff
@@ -303,26 +274,6 @@ let rec mktailpat nilloc = let open Location in function
...
-||||||| a198127529
-let mkexp_type_constraint ?(ghost=false) ~loc e t =
-  let desc =
-    match t with
-  | N_ary.Pconstraint t -> Pexp_constraint(e, t)
-  | N_ary.Pcoerce(t1, t2)  -> Pexp_coerce(e, t1, t2)
-  in
-  if ghost then ghexp ~loc desc
-  else mkexp ~loc desc
-=======
```
Comment (ncik-roberts) [inline ocaml/parsing/parser.mly:325]: (I'm attempting to leave the comment on the portion below the `=======` but GitHub isn't cooperating.)

I find this implementation to be pretty sad. `Pcoerce` just drops `~modes`? Sure, I think it's always empty here, but the code structure doesn't make that clear. Probably we should move the modes to the payload of `Pconstraint` &mdash; that will simplify some other things too. Let's CR this.

(I know you didn't write this code.)

File: ocaml/parsing/parsetree.mli [hunk 4]
```diff
@@ -513,52 +513,6 @@ and function_param_desc =
       - [?l:(P = E0)]
         when [lbl] is {{!Asttypes.arg_label.Optional}[Optional l]}
         and [exp0] is [Some E0]
-<<<<<<< HEAD
```
Comment (ncik-roberts) [inline ocaml/parsing/parsetree.mli:516]: I think there is a merge conflict farther up in this file that remains unresolved.

File: ocaml/typing/typecore.ml [hunk 5]
```diff
@@ -5183,29 +4904,11 @@ let vb_exp_constraint {pvb_expr=expr; pvb_pat=pat; pvb_constraint=ct; pvb_modes=
...
-      let loc = { expr.pexp_loc with Location.loc_ghost = true } in
-      Exp.coerce ~loc expr ground coercion
-  | Some (Pvc_constraint { locally_abstract_univars=vars;typ}) ->
-      let loc_start = pat.ppat_loc.Location.loc_start in
-      let loc = { expr.pexp_loc with loc_start; loc_ghost=true } in
-      let expr = Exp.constraint_ ~loc expr (add_mode_annot_attrs mode_annot_attr typ) in
-      List.fold_right (Exp.newtype ~loc) vars expr
-=======
       Exp.coerce ~loc expr ground coercion |> maybe_add_modes_constraint
   | Some (Pvc_constraint { locally_abstract_univars=vars;typ}) ->
```
Comment (ncik-roberts) [inline ocaml/typing/typecore.ml]: I suggest punning the pattern match on the `locally_abstract_univars` field to match the HEAD side of the diff and to reduce drift with upstream.

**Review by ncik-roberts — APPROVED** [review]
Thanks for doing this. Sorry that it involved quashing drift that I introduced in two copies of the same changeset. I redid the diff resolution and checked my work against yours. It was almost exactly the same. I left a comment in the small handful of places that I felt my resolution might be better.

**Comment by mshinwell** [comment]
See ocaml-flambda/flambda-backend#2972

**Pull request closed without merging**